### PR TITLE
Add sanity check to sync between store and wallet

### PR DESCRIPTION
**Pull request by quantumagi**
Relates to #137 

- On restart we have to check that our transaction's UTXO's are still valid.
- Once ProcessTransaction has been called we want to check that UTXO's are not re-used.

### CODE REVIEW

File: src/Stratis.FederatedPeg.Tests/TransactionBuilderTests.cs [hunk 1]
```diff
@@ -128,43 +121,171 @@ public void BuildTransactionBuildsDeterministicTransactions()
...
+                new Recipient.Recipient
+                {
+                    Amount = Money.COIN * 70,
+                    ScriptPubKey = PayToPubkeyHashTemplate.Instance.GenerateScriptPubKey(new Key().PubKey.Hash)
+                }
+            };
+
+            string opReturnData2 = blockHeight.ToString();
+
+            // Build the multisig transaction template.
```
Comment (monsieurleberre) [inline src/Stratis.FederatedPeg.Tests/TransactionBuilderTests.cs]: Is there a way to get a bit less duplicate code ?

File: src/Stratis.FederatedPeg.Tests/TransactionBuilderTests.cs [hunk 2]
```diff
@@ -128,43 +121,171 @@ public void BuildTransactionBuildsDeterministicTransactions()
...
+            Assert.True(this.SanityCheck(transaction2));
+
+        }
+
+        /// <summary>
+        /// Verifies that the transaction's inout UTXO's have been reserved by the wallet.
+        /// </summary>
+        /// <param name="transaction">The transaction to check.</param>
+        /// <returns><c>True</c> if all's well and <c>false</c> otherwise.</returns>
+        private bool SanityCheck(Transaction transaction)
```
Comment (monsieurleberre) [inline src/Stratis.FederatedPeg.Tests/TransactionBuilderTests.cs]: Is it possible to reuse the 
public async Task SanityCheckAsync() or its private Sync version ?

File: src/Stratis.FederatedPeg.Tests/TransactionBuilderTests.cs [hunk 3]
```diff
@@ -128,43 +121,171 @@ public void BuildTransactionBuildsDeterministicTransactions()
...
+            return true;
+        }
+
+        /// <summary>
+        /// Creates a directory for a test, based on the name of the class containing the test and the name of the test.
+        /// </summary>
+        /// <param name="caller">The calling object, from which we derive the namespace in which the test is contained.</param>
+        /// <param name="callingMethod">The name of the test being executed. A directory with the same name will be created.</param>
+        /// <returns>The path of the directory that was created.</returns>
+        public static string CreateTestDir(object caller, [System.Runtime.CompilerServices.CallerMemberName] string callingMethod = "")
```
Comment (monsieurleberre) [inline src/Stratis.FederatedPeg.Tests/TransactionBuilderTests.cs]: Maybe a new helper in Utils\ folder for those static methods ? no big deal for now

File: src/Stratis.FederatedPeg.Tests/CrossChainTransferStoreTests.cs [hunk 4]
```diff
@@ -92,17 +97,17 @@ public void SynchronizeSynchronizesWithChainAndSurvivesRestart()
         {
             ConcurrentChain chain = BuildChain(5);
 
-            string dataDir = GetTestDirectoryPath(this);
+            string dataDir = CreateTestDir(this);
+            var dataFolder = new DataFolder(dataDir);
             var nodeSettings = new NodeSettings(this.network, NBitcoin.Protocol.ProtocolVersion.ALT_PROTOCOL_VERSION,
```
Comment (monsieurleberre) [inline src/Stratis.FederatedPeg.Tests/CrossChainTransferStoreTests.cs]: As discussed, for now OK, but if we can avoid using NodeSettings and instead use mocked IFederationGatewaySettings then we shouldn't have tests randomly failing

File: src/Stratis.FederatedPeg.Tests/TransactionBuilderTests.cs (hunk 2 above)
Comment (quantumagi) [reply src/Stratis.FederatedPeg.Tests/TransactionBuilderTests.cs]: Done

File: src/Stratis.FederatedPeg.Tests/CrossChainTransferStoreTests.cs (hunk 4 above)
Comment (quantumagi) [reply src/Stratis.FederatedPeg.Tests/CrossChainTransferStoreTests.cs]: Done

**Pull request merged**